scipy/stats/_hypotests.py

```diff
@@ -9,6 +9,223 @@
+.split('\n'))
+_nan_policy_parameter_doc = Parameter(_name, _type, _desc)
+_nan_policy_parameter = inspect.Parameter(_name,
+                                          inspect.Parameter.KEYWORD_ONLY,
+                                          default='propagate')
+
+
+def _vectorize_hypotest_factory(result_object, default_axis=0,
+                                n_samples=1, paired=False,
+                                result_unpacker=None, too_small=0,
+                                nan_policy_position=None, vectorized=False):
+
+    if result_unpacker is None:
+        def result_unpacker(res):
+            return res[..., 0], res[..., 1]
+
+    def is_too_small(samples):
+        for sample in samples:
+            if len(sample) <= too_small:
+                return True
+        return False
+
+    def vectorize_hypotest_decorator(hypotest_fun_in):
+        @wraps(hypotest_fun_in)
+        def vectorize_hypotest_wrapper(*args, axis=default_axis,
+                                       nan_policy='propagate', _no_deco=False,
+                                       **kwds):
+
+            if _no_deco:  # for testing, decorator does nothing
+                return hypotest_fun_in(*args, **kwds)
+
+            if (nan_policy_position is not None
+                    and len(args) > nan_policy_position):
+                # This is for overriding an existing nan_policy implementation.
+                # Consider what to do if both are specified but not the same.
+                # To detect this, should the default value of keyword arg be
+                # None?
+                nan_policy = args[nan_policy_position]
+
+            # if n_samples is None, all args are samples
+            n_samp = len(args) if n_samples is None else n_samples
+
+            # split samples from other arguments
+            samples = [np.atleast_1d(sample) for sample in args[:n_samp]]
+            args = args[n_samp:]
+
+            if axis is None:
+                samples = [sample.ravel() for sample in samples]
+                axis = 0
+
+            # if axis is not needed, just handle nan_policy and return
+            ndims = np.array([sample.ndim for sample in samples])
+            if np.all(ndims <= 1):
+                # Addresses nan_policy == "raise"
+                contains_nans = []
+                for sample in samples:
+                    contains_nan, _ = (
+                        scipy.stats.stats._contains_nan(sample, nan_policy))
+                    contains_nans.append(contains_nan)
+
+                # Addresses nan_policy == "propagate"
+                # Consider adding option to let function propagate nans, but
+                # currently the hypothesis tests this is applied to do not
+                # propagate nans in a sensible way
+                if any(contains_nans) and nan_policy == 'propagate':
+                    return result_object(np.nan, np.nan)
```

Comment:
To generalize to statistics with only one output, probably should pass only `np.nan` into `result_object`, then have `result_object` be responsible for placing it wherever it needs to go. This is probably a bit cleaner than what I'm doing now, which is ignoring on of these `nan`s in the `result_object` of the statistics with only one output (`gmean`).